index.js

```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
-
-                    const ONE_DAY = 1000 * 60 * 60 * 24
-
-                    const upcomingInvoiceDate = upcomingInvoice.created * 1000
-
-                    if (
-                        !upcomingInvoice.error &&
-                        upcomingInvoice.created &&
-                        upcomingInvoiceDate - new Date().getTime() < ONE_DAY * global.invoiceNotificationPeriod &&
-                        upcomingInvoice.amount_due / 100 > global.invoiceAmountThreshold
-                    ) {
-                        posthog.capture('upcoming_invoice', {
-                            stripe_customer_id: customer.id,
-                            invoice_date: new Date(upcomingInvoiceDate).toLocaleDateString('en-GB'),
-                            invoice_current_amount: upcomingInvoice.amount_due / 100,
-                            distinct_id: customer.email || customer.id
-                        })
-                        await cache.set(`last_invoice_${customer.id}`, upcomingInvoiceDate)
-                    }
+                const upcomingInvoiceResponse = await fetchWithRetry(
+                    `https://api.stripe.com/v1/invoices/upcoming?customer=${customer.id}`,
+                    global.defaultHeaders
+                )
+                const upcomingInvoice = await upcomingInvoiceResponse.json()
+
+                const upcomingInvoiceDate = upcomingInvoice.created * 1000
+
+                const invoiceData = {
+                    product: productName,
+                    amount_due: upcomingInvoice.amount_due / 100,
+                    customer: customer.email,
+                    quantity: 0
+                }
+                upcomingInvoice.lines?.data.forEach((line) => {
+                    invoiceData.quantity += line.quantity
+                })
+
+                if (productName in invoicesByProduct) {
+                    invoicesByProduct[productName].push(invoiceData)
+                } else {
+                    invoicesByProduct[productName] = [invoiceData]
+                }
+
+                if (!upcomingInvoice.error && upcomingInvoice.created) {
+                    posthog.capture('Upcoming Invoice', {
+                        amount: invoiceData.amount_due,
+                        invoice_date: new Date(upcomingInvoiceDate).toLocaleDateString('en-GB'),
+                        product: invoiceData.product,
+                        quantity: invoiceData.quantity,
+                        stripe_customer_id: customer.id,
+                        distinct_id: customer.email || customer.id,
+                        $set: customer.email ? { email: customer.email } : undefined
+                    })
                 }
             }
         }
 
-        const customerRecordExists = await storage.get(customer.id)
+        // Note: I think if we want to automatically update users we should just update properties instead of firing events
```

Comment:
Problem is we don't have access to `posthog.people.set` inside plugins yet, so firing an event is actually just a proxy for setting the props.

The alternative is to use the PostHog API, but that adds complexity as it needs a PostHog API key.

Reply:
makes sense – thanks!